pkg/scheduler/plugins/gang/gang_plugin.go

```diff
@@ -0,0 +1,507 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	helpers "k8s.io/component-helpers/scheduling/corev1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	pgLister "sigs.k8s.io/scheduler-plugins/pkg/generated/listers/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingconfig "github.com/koordinator-sh/koordinator/apis/scheduling/config"
+	"github.com/koordinator-sh/koordinator/apis/scheduling/config/validation"
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/frameworkext"
+)
+
+type Status string
+
+const (
+	Name = "Gang"
+
+	//Permit internal status
+	NotFoundInCache Status = "Gang not found in cache"
+	Success         Status = "Success"
+	Wait            Status = "Wait"
+)
+
+var (
+	_ framework.QueueSortPlugin  = &GangPlugin{}
+	_ framework.PreFilterPlugin  = &GangPlugin{}
+	_ framework.PostFilterPlugin = &GangPlugin{}
+	_ framework.ReservePlugin    = &GangPlugin{}
+	_ framework.PostBindPlugin   = &GangPlugin{}
+	_ framework.PostBindPlugin   = &GangPlugin{}
+)
+
+type GangPlugin struct {
+	frameworkHandler framework.Handle
+	podLister        v1.PodLister
+	pgLister         pgLister.PodGroupLister
+	gangCache        *gangCache
+	pluginArgs       *schedulingconfig.GangSchedulingArgs
+}
+
+func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	pluginArgs, ok := args.(*schedulingconfig.GangSchedulingArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type GangSchedulingArgs, got %T", args)
+	}
+	if err := validation.ValidateGangSchedulingArgs(pluginArgs); err != nil {
+		return nil, err
+	}
+
+	gangCache := NewGangCache()
+	gangCache.pluginArgs = pluginArgs
+
+	//podGroup informer
+	extendedHandle, ok := handle.(frameworkext.ExtendedHandle)
+	if !ok {
+		return nil, fmt.Errorf("want handle to be of type frameworkext.ExtendedHandle, got %T", handle)
+	}
+
+	pgInformer := extendedHandle.PodGroupSharedInformerFactory().Scheduling().V1alpha1().PodGroups()
+	pgInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    gangCache.OnGangAdd,
+			UpdateFunc: gangCache.OnGangUpdate,
+			DeleteFunc: gangCache.OnGangDelete,
+		},
+	)
+
+	//pod informer
+	podInformer := extendedHandle.SharedInformerFactory().Core().V1().Pods().Informer()
+	podLister := extendedHandle.SharedInformerFactory().Core().V1().Pods().Lister()
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    gangCache.OnPodAdd,
+			UpdateFunc: gangCache.OnPodUpdate,
+			DeleteFunc: gangCache.OnPodDelete,
+		},
+	)
+
+	gang := &GangPlugin{
+		frameworkHandler: handle,
+		podLister:        podLister,
+		pgLister:         pgInformer.Lister(),
+		gangCache:        gangCache,
+		pluginArgs:       pluginArgs,
+	}
+
+	//recover the gangCache
+	gang.RecoverGangCache()
```

Comment:
RecoverGangCache should not be in new stage. GangCache should also be in framework, after waitForCacheSync Done,  execute framework.GangCache.Recover()

Reply:
solved

Reply:
solved, add gangCache to extendHandler, after running the informer, recover the gang